### DISCUSSION TITLE
[Bitbucket Self Hosted] Onboarding New Option Suggestion

### DISCUSSION THREAD

**Discussion by jerome-nelson** [renovatebot/renovate#12413] [Suggest an Idea]
**Current Issue**

Bitbucket Cloud PRs cannot be `deleted` only `declined`. In this state it's not possible to rename the PR. The only solutions I can think of are to:

1. Switch off onboarding and manually add - which breaks the automated workflow
2. Use `onboardingBranch` or `branchPrefix` to re-trigger, but same issue as above.

**Suggestion**
Add Config Option like `forceOnboarding` - default false. This would override any onboarding re-trigger conditions if set. Renovate wouldn't need to care about any cloud specific changes relating to PR functionality as well.

**Current Options**
![image](https://user-images.githubusercontent.com/1691165/139583154-c187726e-984b-4096-8c12-21a38a617dbc.png)

**Declined State**
![image](https://user-images.githubusercontent.com/1691165/139583181-c67a02fb-8782-44e2-97d4-2451d5d74a83.png)

Thanks for reading :-)

**Comment by rarkins** [comment]
See #10674 for one option

**Comment by jerome-nelson** [comment]
Thanks for the quick answer :-). It still doesn't seem to apply to Bitbucket Cloud. 

(Originally I missed this code)
https://github.com/renovatebot/renovate/blob/main/lib/platform/bitbucket/index.ts#L651 (But it seems like you are already aware of this).

<hr />
I believe in delegating this responsibility to developers - hence the `forceOnboarding: false` .
Alternatively we could check for declined onboarding PRs and check the decline message. 
If this message is `renovate-ignore` then it can be recreated.

I also believe the issue occurs because of `matchState` and the `PrState` type not having platform specific implementations
https://github.com/renovatebot/renovate/blob/569bee6e7d78956bac61831794ba54272f314e42/lib/platform/bitbucket/index.ts#L200

<hr>

**API**
It seems that we are returning - which returns a list of ALL PR States
https://api.bitbucket.org/2.0/repositories/{repo}/pullrequests?state=<false>

What should be requested is:
https://api.bitbucket.org/2.0/repositories/{repo}/pullrequests?state=DECLINED

It is also possible to set a "reason" message on declined PR - this could be set to renovate-ignore and then caught by renovate - in order to stay in line with the Azure fix.

**Screenshots**
API Return Reference
![image](https://user-images.githubusercontent.com/1691165/139599722-8c9a64f7-f375-4300-aaa4-10e5502db701.png)


DEBUG Logging Issues
![image](https://user-images.githubusercontent.com/1691165/139599417-fb72f6a9-e50b-46be-96b2-777c508d8555.png)

**Reply by rarkins** [reply]
> It still doesn't seem to apply to Bitbucket Cloud.

It hasn't been implemented yet, so I'm not sure what you mean? i.e. it doesn't apply to anything at all yet. Do you mean it won't work with Bitbucket Cloud?

**Reply by jerome-nelson** [reply]
I misunderstood the PR state, and yes this solution needs a bitbucket-specific check. I think the solution might be in the comment above somewhere.
Want me to introduce some code changes?

**Reply by rarkins** [reply]
We recommend that you have an agreed issue first before submitting PRs, otherwise there's too much chance we may decline the code. Are you considering the label approach (see https://github.com/renovatebot/renovate/pull/11594) or something different? I just noticed that bitbucket cloud doesn't appear to support labels?

**Reply by jerome-nelson** [reply]
Can transition it - but solutions depend on what you and the rest of the team think:
1) BB Cloud  has a `reason` - option in API for declined PRs, this could be used for `renovate-ignore` approach. That or the PR comments which can also be used.
2) The other issue is the PRState - would recommend making that override-able by platforms
3) Personally I think renovate-ignore is too much config. So would like to discuss that

**Reply by rarkins** [reply]
1a) Is that something people can access via the web UI?
1b) Is reason a freeform field or has a set of allowed values?
1c) We prefer not to use PR comments
2) I don't understand what you mean, please direct me to what you mean specifically
3) It's planned to be zero config, so one of us has a misunderstanding

**Reply by jerome-nelson** [reply]
1a) Yes (screenshots below)
1b) It's a freeform field
1c) Understood
2) https://github.com/renovatebot/renovate/blob/main/lib/types/pr-state.ts - This is different for BB Cloud
3) It's more of a personal preference; it's not important or relevant in hindsight :-) 


![image](https://user-images.githubusercontent.com/1691165/139694636-6a427774-6f54-466b-926d-e4648fcfb1d4.png)
![image](https://user-images.githubusercontent.com/1691165/139694689-23c2eecd-81f3-4bf1-8175-1a7b0839ef56.png)

**Reply by rarkins** [reply]
(1) Can the reason for declining be edited later, or it's not possible to change once it's set at time of Decline?
(2) Each platform maps their own values to these, forget about it for now

**Reply by jerome-nelson** [reply]
Sorry for the late reply
1) No.
2) Could you help me find the files for this?

**Reply by rarkins** [reply]
1. If it can't be changed later, then it's not right for this feature

**Reply by jerome-nelson** [reply]
Understood - but in BB Cloud, when you decline a PR, it becomes immutable. Instead of // renovate-ignore approach, how about simply checking if BB PR is in declined state? 
Once declined it cannot be re-opened

**Reply by rarkins** [reply]
People usually decline PRs because they want them ignored, not because they want another one created.

Can you describe again what your use case is? e.g. You have an onboarding PR which you edited and now you want to start over?

**Reply by jerome-nelson** [reply]
Sure no problem :-) hope this makes it clearer

**Use Case**
A user accidentally declines a PR on BB Cloud, not knowing that this Onboarding PR cannot be recreated. The message on the PR doesn't take into account that Bitbucket PRs are not change-able once declined.
<hr>

I simply made a mistake, whilst testing. I was new to Renovate and when I look at https://docs.renovatebot.com/getting-started/installing-onboarding/ it doesn't mention anything about Declined PRs.

If I want to use this in automation and then a team member in my BB accidentally declines a branch it creates more work

**Reply by rarkins** [reply]
I don't think we can change the current behavior because most people decline the PR intentionally.

**Reply by jerome-nelson** [reply]
The change isn't the issue - it would be nice to have. The issue is the lack of clear guidelines - so a simpler solution: 
1. We edit the BB Cloud Onboarding Comment to inform new users not to decline PRs
2. Update the Docs https://docs.renovatebot.com/getting-started/installing-onboarding/ to add some kind of warning for BB Cloud users.

Please bear in mind, I mention Onboarding, but any PR which is declined on BB Cloud will have this issue

**Comment by tbugfinder** [comment]
I'd also like to add that the Bitbucket environment requires some sort of force recreating the onboarding pullrequest.
A PR decline might be intentionally at some point in time however one year later recreating onboarding PR might be the  *intented* target but not possible as tooling is too strict.
As mentioned before, first renovate steps might also end up in declining the PR without "knowing" the logic that the declined PRs have side-effects.
Unfortunately work on #11594 stopped.
